core/azure-core/src/main/java/com/azure/core/polling/Poller.java

```diff
@@ -0,0 +1,240 @@
+     * @param cancelOperation cancel operation
+     */
+    public Poller(PollerOptions pollerOptions, Function<PollResponse<T>, PollResponse<T>> pollOperation, Consumer<Poller> cancelOperation) {
+        this(pollerOptions, pollOperation);
+        this.cancelOperation = cancelOperation;
+    }
+
+    /*
+     * We will maintain single instance of fluxHandle for one poller.
+     */
+    private Flux<PollResponse<T>> createFlux() {
+        if (fluxHandle == null) {
+            fluxHandle = sendPollRequestWithDelay()
+                .flux()
+                .repeat()
+                .takeUntil(pollResponse -> needsPolling(pollResponse));
+        }
+        return fluxHandle;
+    }
+
+    private boolean needsPolling(PollResponse<T> pollResponse) {
+        return (  pollResponse.isDone());
+    }
+
+    /**
+     * This will cancel polling from Azure Service if supported by service
+     * @throws UnsupportedOperationException when cancel operation is not suported.
+     */
+    public void cancelOperation() throws UnsupportedOperationException {
+        if (cancelOperation == null) {
+            throw new UnsupportedOperationException("Cancel operation is not supported on this service/resource.");
+        }
+
+        // We can not cancel an operation if it was never started
+        // or it is in its terminal state.
+        // Check cancelInitiated: to protect against multiple time call to cancel operation
+        if (cancelInitiated || (pollResponse != null && pollResponse.getStatus() != PollResponse.OperationStatus.IN_PROGRESS)) {
+            return;
+        }
+        //Time to call cancel
+        cancelOperation.accept(this);
+        cancelInitiated = true;
+    }
+
+    /**
+     * Enable client to subscribe and receive all the responses.
+     * @return Return poll response as Flux
+     */
+    public Flux<PollResponse<T>> getObserver() {
+         createFlux();
+         return fluxHandle;
+    }
+
+    /**
+     * Calls poll operation once in sync.
+     * @return  Mono of poll response
+     */
+    public Mono<PollResponse<T>> poll() {
+        updatePollOperationSync();
+        return Mono.just(pollResponse);
+    }
+
+    /**
+     * This will block till poll operation is complete
+     * @return returns poll response
+     */
+    public PollResponse<T> block() {
+        return getObserver().blockLast();
+    }
+
+    /*
+     * Calls poll operation function and update pollResponse.
+     */
+    private  void updatePollOperationSync() {
+        pollResponse = pollOperation.apply(pollResponse);
+    }
+
+    /*
+     * Get whether or not this PollStrategy's long running operation is done.
+     * @return Whether or not this PollStrategy's long running operation is done.
+     */
+    private Mono<PollResponse<T>> sendPollRequestWithDelay() {
+        return Mono.defer(() -> delayAsync().then(Mono.defer(() -> {
+            if (!isTerminalState()) {
+                updatePollOperationSync();
```

Comment:
Another example where `uploadPollOperation()` method should be async. You cannot block a reactor. For now, if you want to keep everything sync, you can change the 2nd `Mono.defer()` to `Mono.fromCallable()`.